[MISC] Added PID controller util, quadcopter PID controller, flight examples

jebbrysacz opened the pull request:
Abstracting the need for individual rotor control for quadcopter through a pseudo-nested PID controller. Makes flying to a target or targets simple without the need for reinforcement learning. Example flight to targets is in `examples/fly_route.py`. Note that the parameters are tuned for flight, but not optimized.

Run with: `python fly_route.py`.

zhouxian commented:
could you move the pid controller into your example?
if we want to add a independent pid controller class, maybe we need to design more carefully

KafuuChikai commented:
Hi @jebbrysacz ,

We noticed that you've implemented a basic **PID controller** for your drone simulations. That's fantastic! 😀 We're also working on something similar and are looking to make it a bit more complex.

We're considering adding more **realistic actuators**, **mixers**, and advanced PID controllers with features like **angular velocity loops**, **attitude loops**, **integral saturation**, and even **differentiable controllers**. We're planning to migrate the work from [Omnidrone](https://github.com/btx0424/OmniDrones/tree/main) to Genesis.

We'd love to hear your thoughts on this and learn about your plans. Perhaps there's an opportunity for us to collaborate and combine our efforts.

Looking forward to hearing from you!

Best regards,
Xian Wang

jebbrysacz commented:
> could you move the pid controller into your example? if we want to add a independent pid controller class, maybe we need to design more carefully

Zhou, I've just moved the pid controller to the examples folder and deleted the file in utils. Please let me know if there's anything else.

jebbrysacz commented:
> Hi @jebbrysacz ,
> 
> We noticed that you've implemented a basic **PID controller** for your drone simulations. That's fantastic! 😀 We're also working on something similar and are looking to make it a bit more complex.
> 
> We're considering adding more **realistic actuators**, **mixers**, and advanced PID controllers with features like **angular velocity loops**, **attitude loops**, **integral saturation**, and even **differentiable controllers**. We're planning to migrate the work from [Omnidrone](https://github.com/btx0424/OmniDrones/tree/main) to Genesis.
> 
> We'd love to hear your thoughts on this and learn about your plans. Perhaps there's an opportunity for us to collaborate and combine our efforts.
> 
> Looking forward to hearing from you!
> 
> Best regards, Xian Wang

Hello Xian @KafuuChikai ! 

I was also planning to make my drone controller a bit more complex. I was first thinking of adding in nested pid capabilities for angular velocity and attitude to the original controller, although this may conflict with the way the simulation handles time stepping. After that I was planning to build a more complex control algorithm class, maybe model predictive control. 

It sounds like you are planning to add in many more drone control features. I think a lot of the work I would like to accomplish aligns with what you all are doing. I'd be very excited to collaborate with you!! If you'd like, you can contact me at my email: jebbrysacz@gmail.com. I look forward to hearing back!!

Best,
Jeb Brysacz

KafuuChikai commented:
> > Hi @jebbrysacz ,
> > We noticed that you've implemented a basic **PID controller** for your drone simulations. That's fantastic! 😀 We're also working on something similar and are looking to make it a bit more complex.
> > We're considering adding more **realistic actuators**, **mixers**, and advanced PID controllers with features like **angular velocity loops**, **attitude loops**, **integral saturation**, and even **differentiable controllers**. We're planning to migrate the work from [Omnidrone](https://github.com/btx0424/OmniDrones/tree/main) to Genesis.
> > We'd love to hear your thoughts on this and learn about your plans. Perhaps there's an opportunity for us to collaborate and combine our efforts.
> > Looking forward to hearing from you!
> > Best regards, Xian Wang
> 
> Hello Xian @KafuuChikai !
> 
> I was also planning to make my drone controller a bit more complex. I was first thinking of adding in nested pid capabilities for angular velocity and attitude to the original controller, although this may conflict with the way the simulation handles time stepping. After that I was planning to build a more complex control algorithm class, maybe model predictive control.
> 
> It sounds like you are planning to add in many more drone control features. I think a lot of the work I would like to accomplish aligns with what you all are doing. I'd be very excited to collaborate with you!! If you'd like, you can contact me at my email: [jebbrysacz@gmail.com](mailto:jebbrysacz@gmail.com). I look forward to hearing back!!
> 
> Best, Jeb Brysacz

Hi Jeb,

Thank you for your interest in collaborating! We would be delighted to work with you and will be reaching out to you soon. In the meantime, please feel free to contact me at my email: xianwang@zju.edu.cn.

Best regards,
Xian Wang

YilingQiao commented:
Thanks for your contribution!
